pkg/reconciler/namespace/namespace.go

```diff
@@ -0,0 +1,191 @@
+	defaultDomain := cfg.Domain.LookupDomainForLabels(nil)
+	if defaultDomain == "" {
+		c.Logger.Infof("Skipping wildcard certificate creation for namespace %s. No default domain defined.", ns.Name)
+		return nil
+	}
+
+	domain, err := getWildcardDomain(cfg.Network.DomainTemplate, defaultDomain, ns.Name)
+	if err != nil {
+		c.Logger.Errorf("Failed to apply domain template %s to domain %s and namespace %s", cfg.Network.DomainTemplate, defaultDomain, ns.Name)
+	}
+
+	desiredCert := makeDesiredCertificate(ns, []string{domain})
+
+	certs, err := c.ServingClientSet.NetworkingV1alpha1().Certificates(ns.Name).List(metav1.ListOptions{})
+	if err != nil {
+		return err
+	}
+	existingCert, err := getExistingCert(domain, certs)
+
+	if apierrs.IsNotFound(err) {
+		cert, err := c.ServingClientSet.NetworkingV1alpha1().Certificates(ns.Name).Create(desiredCert)
+		if err != nil {
+			c.Logger.Errorw("Failed to create namespace certificate", zap.Error(err))
+			c.Recorder.Eventf(ns, corev1.EventTypeWarning, "Creation Failed",
+				"Failed to create Knative certificate %s/%s: %v", ns.Name, desiredCert.ObjectMeta.Name, err)
+			return nil
+		}
+		c.Recorder.Eventf(cert, corev1.EventTypeNormal, "Created",
+			"Created Knative Certificate %s/%s", ns.Name, cert.ObjectMeta.Name)
+	} else if err != nil {
+		return err
+	} else if !equality.Semantic.DeepEqual(existingCert.Spec, desiredCert.Spec) {
+		copy := existingCert.DeepCopy()
+		copy.Spec = desiredCert.Spec
+		_, err := c.ServingClientSet.NetworkingV1alpha1().Certificates(copy.Namespace).Update(copy)
+		if err != nil {
+			c.Logger.Errorw("Failed to update Knative Certificate", zap.Error(err))
+			c.Recorder.Eventf(existingCert, corev1.EventTypeWarning, "UpdateFailed",
+				"Failed to update Knative Certificate %s/%s: %v", desiredCert.Namespace, desiredCert.Name, err)
+			return err
+		}
+		c.Recorder.Eventf(existingCert, corev1.EventTypeNormal, "Updated",
+			"Updated Spec for Knative Certificate %s/%s", desiredCert.Namespace, desiredCert.Name)
+		return nil
+	}
+
+	return nil
+}
+
+func makeDesiredCertificate(namespace *corev1.Namespace, dnsNames []string) *v1alpha1.Certificate {
+	name := wildcardName(namespace.Name)
+	return &v1alpha1.Certificate{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: namespace.Name,
+		},
+		Spec: v1alpha1.CertificateSpec{
+			DNSNames:   dnsNames,
+			SecretName: name,
+		},
+	}
+}
+
+func getWildcardDomain(tmpl, domain, namespace string) (string, error) {
+	data := network.DomainTemplateValues{
+		Name:      "*",
+		Domain:    domain,
+		Namespace: namespace,
+	}
+	buf := bytes.Buffer{}
+	t, err := template.New("domain-template").Parse(tmpl)
+
+	if err != nil {
+		return "", err
+	}
+	err = t.Execute(&buf, data)
+
+	if err != nil {
+		return "", err
+	}
+
+	return buf.String(), nil
+}
+
+func getExistingCert(domain string, certs *v1alpha1.CertificateList) (*v1alpha1.Certificate, error) {
+	testDomain := strings.SplitAfterN(domain, ".", 2)[1]
```

Comment:
Are you guaranteed to have at least 2 elements here? Same below.
Also I'd rename `testDomain` to `q` or `query`.

Reply:
We're not guaranteed since a DomainTemplate value of `{{.Domain}}` seems to be acceptable. However, if I validate that value as suggested by @ZhiminXiang  then it won't be an issue. For the second instance I don't think any CA will issue a cert for anything but an FQDN.